File: Graphs/1073-number-of-enclaves/number-of-enclaves.py
class Solution:
    def dfs(self, vis, grid, sr, sc, n, m):
        if sr < 0 or sr >= n or sc < 0 or sc >= m or vis[sr][sc] != 0 or grid[sr][sc] == 0:
            return

        vis[sr][sc] = 1

        self.dfs(vis, grid, sr - 1, sc, n, m)
        self.dfs(vis, grid, sr + 1, sc, n, m)
        self.dfs(vis, grid, sr, sc - 1, n, m)
        self.dfs(vis, grid, sr, sc + 1, n, m)

    def numEnclaves(self, grid):
        n = len(grid)
        m = len(grid[0])
        vis = [[0 for _ in range(m)] for _ in range(n)]

        # Mark boundary-connected land cells
        for i in range(m):
            if grid[0][i] == 1:
                self.dfs(vis, grid, 0, i, n, m)
            if grid[n - 1][i] == 1:
                self.dfs(vis, grid, n - 1, i, n, m)

        for i in range(n):
            if grid[i][0] == 1:
                self.dfs(vis, grid, i, 0, n, m)
            if grid[i][m - 1] == 1:
                self.dfs(vis, grid, i, m - 1, n, m)

        # Count remaining land cells not visited
        ans = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1 and vis[i][j] == 0:
                    ans += 1

        return ans

File: Graphs/1073-number-of-enclaves/number-of-enclaves.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& vis, vector<vector<int>>& grid, int sr, int sc, int n, int m){
        if(sc < 0 || sc >= m || sr < 0 || sr >= n || vis[sr][sc] != 0 || grid[sr][sc] == 0)
            return;
        
        vis[sr][sc] = 1;

        dfs(vis, grid, sr - 1, sc, n, m);
        dfs(vis, grid, sr + 1, sc, n, m);
        dfs(vis, grid, sr, sc - 1, n, m);
        dfs(vis, grid, sr, sc + 1, n, m);
    }

    int numEnclaves(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));

        for(int i = 0; i < m; i++){
            if(grid[0][i] == 1)
                dfs(vis, grid, 0, i, n,  m);
            
            if(grid[n-1][i] == 1)
                dfs(vis, grid, n-1, i, n,  m);
        }

        for(int i = 0; i < n; i++){
            if(grid[i][0] == 1)
                dfs(vis, grid, i, 0, n,  m);
            
            if(grid[i][m-1] == 1)
                dfs(vis, grid, i, m-1, n,  m);
        }

        int ans = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1 && vis[i][j] == 0)
                    ans++;
            }
        }

        return ans;
    }
};